File: server/src/routes/review.js
import { Router } from 'express';
import { reviewCode } from '../services/gemini.js';
import { fetchGithubFiles } from '../services/github.js';
import { optionalAuth } from '../middleware/auth.js';
import { saveReview } from './reviews.js';

const router = Router();

// Wrap async handlers so thrown errors reach the error middleware.
const wrap = (fn) => (req, res, next) => Promise.resolve(fn(req, res, next)).catch(next);

/**
 * POST /api/review
 * Body: { code, language?, filename?, source? }
 * Reviews a single snippet pasted or uploaded by the user.
 * If the request is authenticated, the review is saved to the user's history.
 */
router.post(
  '/review',
  optionalAuth,
  wrap(async (req, res) => {
    const { code, language, filename, source } = req.body || {};
    if (!code || !code.trim()) {
      return res.status(400).json({ error: 'Field "code" is required.' });
    }
    const review = await reviewCode({ code, language, filename });
    const savedId = await saveReview({
      userId: req.userId,
      source: source === 'upload' ? 'upload' : 'paste',
      filename,
      review,
    });
    res.json({ filename: filename || null, review, savedId });
  })
);

/**
 * POST /api/review/github
 * Body: { url }
 * Fetches code from a public GitHub repo / PR / file URL and reviews each file.
 */
router.post(
  '/review/github',
  optionalAuth,
  wrap(async (req, res) => {
    const { url } = req.body || {};
    if (!url || !url.trim()) {
      return res.status(400).json({ error: 'Field "url" is required.' });
    }

    const { source, files } = await fetchGithubFiles(url);

    // Review each fetched file. Failures on one file shouldn't sink the others.
    const reviews = [];
    for (const file of files) {
      try {
        const review = await reviewCode({
          code: file.content,
          language: 'auto',
          filename: file.path,
        });
        const savedId = await saveReview({
          userId: req.userId,
          source: 'github',
          filename: file.path,
          review,
        });
        reviews.push({ filename: file.path, review, savedId });
      } catch (err) {
        reviews.push({ filename: file.path, error: err.message });
      }
    }

    res.json({ source, fileCount: files.length, reviews });
  })
);

export default router;

File: server/app/routers/review.py
"""Code-review routes: paste/upload snippet and GitHub URL."""
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from ..auth import optional_user
from ..gemini_service import review_code
from ..github_service import fetch_github_files
from .reviews import save_review

router = APIRouter()


class ReviewIn(BaseModel):
    code: str | None = None
    language: str | None = "auto"
    filename: str | None = ""
    source: str | None = "paste"


class GithubIn(BaseModel):
    url: str | None = None


@router.post("/review")
def post_review(body: ReviewIn, user_id=Depends(optional_user)):
    """Review a single snippet. If authenticated, save it to the user's history."""
    if not body.code or not body.code.strip():
        raise HTTPException(400, 'Field "code" is required.')
    review = review_code(body.code, body.language or "auto", body.filename or "")
    saved_id = save_review(
        user_id, "upload" if body.source == "upload" else "paste", body.filename, review
    )
    return {"filename": body.filename or None, "review": review, "savedId": saved_id}


@router.post("/review/github")
def post_github(body: GithubIn, user_id=Depends(optional_user)):
    """Fetch code from a public GitHub repo / PR / file URL and review each file."""
    if not body.url or not body.url.strip():
        raise HTTPException(400, 'Field "url" is required.')

    result = fetch_github_files(body.url)
    source, files = result["source"], result["files"]

    reviews = []
    for f in files:
        try:
            review = review_code(f["content"], "auto", f["path"])
            saved_id = save_review(user_id, "github", f["path"], review)
            reviews.append({"filename": f["path"], "review": review, "savedId": saved_id})
        except Exception as err:  # noqa: BLE001
            reviews.append({"filename": f["path"], "error": str(err)})

    return {"source": source, "fileCount": len(files), "reviews": reviews}
